src/Nethermind/Nethermind.Core/Utils/WriteBatcher.cs

```diff
@@ -0,0 +1,73 @@
+// SPDX-FileCopyrightText: 2024 Demerzel Solutions Limited
+// SPDX-License-Identifier: LGPL-3.0-only
+
+using System;
+using System.Collections.Concurrent;
+using System.Threading;
+
+namespace Nethermind.Core.Utils;
+
+/// <summary>
+/// Batches writes into a set of concurrent batches. For cases where throughput matter, but not atomicity.
+/// </summary>
+public class WriteBatcher : IWriteBatch
+{
+    private long _counter = 0;
+    private readonly ConcurrentQueue<IWriteBatch> _batches = new();
+    private readonly IKeyValueStoreWithBatching _underlyingDb;
+    private bool _disposing = false;
+
+    public WriteBatcher(IKeyValueStoreWithBatching underlyingDb)
+    {
+        _underlyingDb = underlyingDb;
+    }
+
+    public void Dispose()
+    {
+        _disposing = true;
+        foreach (IWriteBatch batch in _batches)
+        {
+            batch.Dispose();
+        }
+    }
+
+    public void PutSpan(ReadOnlySpan<byte> key, ReadOnlySpan<byte> value, WriteFlags flags = WriteFlags.None)
+    {
+        if (_disposing) throw new InvalidOperationException("Trying to set while disposing");
+        if (!_batches.TryDequeue(out IWriteBatch? currentBatch))
+        {
+            currentBatch = _underlyingDb.StartWriteBatch();
+        }
+
+        currentBatch.PutSpan(key, value, flags);
+        long val = Interlocked.Increment(ref _counter);
+        if (val % 10000 == 0)
+        {
+            currentBatch.Dispose();
```

Comment:
potential concurrency issue -> Can you be disposing while other thread calls `PutSpan`? Do we need `_counter < limit` somewhere before `PutSpan`?

Maybe not as ConcurrentQueue would handle it?

Reply:
Yes, that is why the concurrent queue is there.

The return of `Interlocked.Increment` should be atomic so it should persist it eventually. Its a bit random though.

Reply:
If not, dispose should dequeue everything.